{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 26,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gEfWrxaGkfcU",
        "outputId": "339835b5-d532-4ec4-9ddf-ee544d54873f"
      },
      "outputs": [],
      "source": [
        "from reportlab.lib.pagesizes import letter, A4\n",
        "from reportlab.platypus import BaseDocTemplate, SimpleDocTemplate, Paragraph, Spacer, Image, PageBreak, Table, TableStyle, PageTemplate, Frame, Flowable\n",
        "from reportlab.lib.styles import ParagraphStyle, getSampleStyleSheet\n",
        "from reportlab.lib import utils, colors\n",
        "from reportlab.lib.units import inch\n",
        "from reportlab.lib.colors import Color, HexColor\n",
        " \n",
        "from reportlab.lib import colors\n",
        "from reportlab.platypus.flowables import KeepTogether\n",
        "from reportlab.pdfbase.pdfmetrics import stringWidth\n",
        " \n",
        "import json\n",
        "import os\n",
        " \n",
        "# Custom page size and margins\n",
        "PAGE_WIDTH, PAGE_HEIGHT = A4\n",
        "TOP_MARGIN = 1.7 * inch\n",
        "BOTTOM_MARGIN = 0.29 * inch\n",
        "LEFT_MARGIN = 0.5 * inch\n",
        "RIGHT_MARGIN = 0.5 * inch\n",
        " \n",
        "TITLE_TEXT = \"Rare Disease Profiling\"\n",
        "TITLE_FONT = \"Helvetica-Bold\"\n",
        "TITLE_FONT_SIZE = 16\n",
        " \n",
        "TITLE_WIDTH = stringWidth(TITLE_TEXT, TITLE_FONT, TITLE_FONT_SIZE)\n",
        "TITLE_HEIGHT = TITLE_FONT_SIZE\n",
        " \n",
        "AVAILABLE_PAGE_WIDTH = PAGE_WIDTH - LEFT_MARGIN - RIGHT_MARGIN\n",
        " \n",
        "TITLE_PADDING = 10\n",
        "PATIENT_INFO_PADDING = 5\n",
        " \n",
        " \n",
        "styles = getSampleStyleSheet()\n",
        "#justified_paragraph = styles[\"Normal\"]\n",
        "#justified_paragraph.alignment = 4  # 4 corresponds to 'JUSTIFY'\n",
        " \n",
        "#center_paragraph = styles[\"Normal\"]\n",
        "#center_paragraph.alignment = 1  # 1 corresponds to 'CENTER'\n",
        " \n",
        "# Create a new style object for justified paragraph based on the existing Normal style\n",
        "justified_paragraph = ParagraphStyle(\"Justified\", parent=styles[\"Normal\"])\n",
        "justified_paragraph.alignment = 4  # 4 corresponds to 'JUSTIFY'\n",
        " \n",
        "center_paragraph = ParagraphStyle(\"Centered\", parent=styles[\"Normal\"])\n",
        "center_paragraph.alignment = 1  # 4 corresponds to 'JUSTIFY'\n",
        " \n",
        " \n",
        "# Create a new style object for center-aligned paragraph based on the existing Normal style\n",
        "imageTitle_style = ParagraphStyle(\"Centered\", parent=styles[\"Normal\"])\n",
        "imageTitle_style.alignment = 1  # 1 corresponds to 'CENTER'\n",
        "imageTitle_style.fontSize = 9\n",
        " \n",
        "# Create a new style object for center-aligned paragraph based on the existing Normal style\n",
        "imageCaption_style = ParagraphStyle(\"Centered\", parent=styles[\"Normal\"])\n",
        "imageCaption_style.alignment = 1  # 1 corresponds to 'CENTER'\n",
        "imageCaption_style.fontSize = 7\n",
        "imageCaption_style.fontWeight = 'Normal'  # Set font weight to normal\n",
        " \n",
        "# Create a new style object for center-aligned paragraph based on the existing Normal style\n",
        "title_style = ParagraphStyle(\"Centered\", parent=styles[\"Heading2\"])\n",
        "title_style.alignment = 1  # 1 corresponds to 'CENTER'\n",
        "title_style.fontSize = 15\n",
        " \n",
        "# Create a new style object for center-aligned paragraph based on the existing Normal style\n",
        "left_paragraph = ParagraphStyle(\"Left\", parent=styles[\"Normal\"])\n",
        "left_paragraph.alignment = 0  # 1 corresponds to 'CENTER'\n",
        "\n",
        "right_paragraph = ParagraphStyle(\"Right\", parent=styles[\"Normal\"])\n",
        "right_paragraph.alignment = 2  # 1 corresponds to 'CENTER'\n",
        "\n",
        " \n",
        "def json_to_pdf(patient_metadata_json, result_json, pdf_file_path):\n",
        "    # Open the JSON file\n",
        "    with open(patient_metadata_json, 'r') as f:\n",
        "        # Load the JSON data\n",
        "        metadata = json.load(f)\n",
        "    metadata = metadata['metadata']\n",
        " \n",
        "    with open(result_json, 'r') as f:\n",
        "        # Load the JSON data\n",
        "        result = json.load(f)\n",
        "    result = result['condition']\n",
        "    # Patient Information\n",
        " \n",
        "    patient_id_num = metadata[\"Patient ID\"]\n",
        "    patient_ethnicity = metadata[\"Ethnicity\"]\n",
        " \n",
        "    patient_id_data = [\n",
        "        [Paragraph(\"Name:\", left_paragraph), Paragraph(\":  \"+metadata[\"Name\"],left_paragraph), Paragraph(f\"Patient ID\",left_paragraph),Paragraph(\":  \"+metadata[\"Patient ID\"],left_paragraph)],\n",
        "        [Paragraph(\"Gender\", left_paragraph), Paragraph(\":  \"+metadata[\"Gender\"],left_paragraph), Paragraph(f\"Ethnicity\",left_paragraph),Paragraph(\":  \"+metadata[\"Ethnicity\"],left_paragraph)],\n",
        "        [Paragraph(\"Date of Birth\", left_paragraph), Paragraph(\":  \"+str(metadata[\"Date of Birth\"]),left_paragraph), None,None],\n",
        "    ]\n",
        "    patient_id_table_colWidths = [100, 250, 100, 100]\n",
        "    patient_id_table = Table(patient_id_data, colWidths=patient_id_table_colWidths)\n",
        "    patient_id_table.setStyle([\n",
        "        ('VALIGN', (0, 0), (-1, -1), 'TOP'),  # Align images at the top of the cells\n",
        "        ('ALIGN', (0, 0), (-1, -1), 'LEFT'),  # Align images to the left\n",
        "        ('BOTTOMPADDING', (0, 0), (-1, -1), 2),  # Padding at the bottom of the cells\n",
        "        ('TOPPADDING', (0, 0), (-1, -1), 2),  # Padding at the top of the cells\n",
        "        ('LEFTPADDING', (0, 0), (-1, -1), 2),  # Padding on the left of the cells\n",
        "        ('RIGHTPADDING', (0, 0), (-1, -1), 2),\n",
        "    ])\n",
        "   \n",
        "    required_width, required_height = patient_id_table.wrap(PAGE_WIDTH, PAGE_HEIGHT)\n",
        "   \n",
        "    def myFirstPage(canvas, doc, table_data=patient_id_table):\n",
        "        canvas.saveState()\n",
        "        # Add the title for the first page\n",
        "        canvas.setFont(TITLE_FONT, TITLE_FONT_SIZE)\n",
        "        canvas.drawString((PAGE_WIDTH/2)-(TITLE_WIDTH/2), PAGE_HEIGHT-TOP_MARGIN-TITLE_HEIGHT, TITLE_TEXT)\n",
        "        canvas.restoreState()\n",
        " \n",
        "        table = table_data\n",
        "       \n",
        "        # Before positioning and drawing the table, calculate its required size\n",
        "        required_width, required_height = table.wrap(PAGE_WIDTH, PAGE_HEIGHT)\n",
        " \n",
        "        table_x_position = LEFT_MARGIN\n",
        "        table_y_position = PAGE_HEIGHT - TOP_MARGIN - required_height - TITLE_HEIGHT - TITLE_PADDING # Adjust based on the table's height\n",
        " \n",
        "        table.wrapOn(canvas, *doc.pagesize)\n",
        "        table.drawOn(canvas, table_x_position, table_y_position)  # Adjust the coordinates as needed\n",
        " \n",
        "    def myLaterPages(canvas, doc, table_data=patient_id_table):\n",
        "        canvas.saveState()\n",
        "        table = table_data\n",
        " \n",
        "        required_width, required_height = table.wrap(PAGE_WIDTH, PAGE_HEIGHT)\n",
        "        globals()['required_height'] = required_height\n",
        " \n",
        "        table_x_position = LEFT_MARGIN\n",
        "        table_y_position = PAGE_HEIGHT - TOP_MARGIN - required_height  # Adjust based on the table's height\n",
        " \n",
        "        table.wrapOn(canvas, *doc.pagesize)\n",
        "        table.drawOn(canvas, table_x_position, table_y_position)  # Adjust the coordinates as needed\n",
        " \n",
        "        canvas.restoreState()\n",
        " \n",
        " \n",
        "    # Create a PDF document with ReportLab\n",
        "    doc = SimpleDocTemplate(pdf_file_path, pagesize=A4, leftMargin=LEFT_MARGIN, rightMargin=RIGHT_MARGIN, topMargin=TOP_MARGIN+required_height+PATIENT_INFO_PADDING, bottomMargin=BOTTOM_MARGIN)\n",
        "    story = []\n",
        " \n",
        " \n",
        "    styles = getSampleStyleSheet()\n",
        " \n",
        " \n",
        "    story.append(Spacer(1, TITLE_HEIGHT))\n",
        "    story.append(Spacer(1, TITLE_PADDING))\n",
        "    story.append(Spacer(1, PATIENT_INFO_PADDING))\n",
        "    story.append(Paragraph(\"Family History\", styles['Heading3']))\n",
        "    story.append(Paragraph(metadata[\"Family History\"], justified_paragraph))\n",
        "    story.append(Paragraph(\"<br/>\", styles['Normal']))  # Add some spacing\n",
        "    story.append(Paragraph(\"Listed of Disease/Condition in this Report\", styles['Heading3']))\n",
        "    for condition, variants in result.items():\n",
        "        story.append(Paragraph(f'---  {condition}', justified_paragraph))\n",
        "    story.append(PageBreak())\n",
        " \n",
        " \n",
        "    story.append(Paragraph(\"Report Disclaimer\", styles['Heading3']))\n",
        "    story.append(Paragraph('''\n",
        "                            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris vehicula libero at sollicitudin dignissim. Nam aliquam arcu ut risus tempus facilisis. Etiam congue tempus tellus. Nulla eu enim a mi sagittis sagittis nec bibendum quam. Proin eget pulvinar tortor. Fusce nibh purus, tempus vitae hendrerit id, pretium sit amet eros. Cras eget dui justo. Ut interdum pulvinar facilisis. Quisque luctus efficitur nisi eu imperdiet. Aliquam fermentum enim non elit aliquet eleifend. Sed nunc sem, tempor in venenatis in, volutpat non magna. Suspendisse sodales, purus nec imperdiet molestie, tortor magna tristique nisi, id pulvinar eros lorem quis lacus. Cras eu augue lorem.\n",
        " \n",
        "                            Pellentesque at turpis eget nunc sollicitudin semper sed at dolor. Nulla tristique maximus tellus at finibus. Vestibulum eu erat sit amet enim tristique condimentum at sed erat. Ut massa quam, euismod non dictum ut, vehicula quis quam. Nullam id lorem maximus, imperdiet neque eget, tristique mauris. Vivamus accumsan mollis lacus et scelerisque. Suspendisse elementum ex sed dui interdum gravida. Fusce vel risus orci. Fusce efficitur, odio id luctus placerat, metus orci sollicitudin mauris, a sagittis urna nulla facilisis nibh. Maecenas luctus semper libero, quis lobortis ligula placerat eu. Sed commodo orci lectus, a luctus tellus lobortis a.\n",
        " \n",
        "                            Duis sit amet tempus metus. Proin semper ac leo et aliquet. Vestibulum porttitor et orci ut commodo. Praesent sed bibendum arcu. Integer rhoncus magna interdum, rutrum augue vel, rutrum turpis. Nulla ut nisi ornare, viverra nisi ut, egestas ipsum. Nam porttitor risus id lorem elementum porttitor. Curabitur sollicitudin metus non euismod ullamcorper. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Phasellus neque nulla, congue sed consequat ac, elementum id nulla. Mauris nec porttitor purus. Mauris dapibus eget magna et tempus. Morbi tincidunt, sapien vel mattis blandit, elit nulla sodales mi, quis semper ante eros vitae nunc. Phasellus euismod diam id urna pulvinar condimentum id non lacus. Ut pretium facilisis tortor.\n",
        " \n",
        "                            Sed eget orci fringilla tortor faucibus luctus euismod quis velit. Suspendisse lorem diam, pellentesque sed sollicitudin at, semper vel leo. Proin id luctus risus. Donec accumsan odio quis nunc ultricies, sed tincidunt metus porta. Vestibulum interdum dui lectus, in placerat nisi tempor quis. Maecenas at erat in nunc tincidunt faucibus eget vitae mauris. Integer ultricies tristique dui non convallis. Curabitur tempor vehicula lacus, sit amet fringilla nisl efficitur a. Fusce vitae faucibus risus. Curabitur ut nulla eget sapien fermentum eleifend. Proin ante magna, pharetra ac tempor ac, lobortis in erat. Nam ipsum tortor, venenatis semper ligula ut, pharetra finibus massa. Fusce euismod vulputate tellus vitae pellentesque. Fusce sit amet tincidunt mi. Aliquam erat volutpat.\n",
        " \n",
        "                            Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Mauris vel semper metus. Sed tortor metus, aliquam non eleifend at, convallis id ligula. Nam fringilla, est non mollis semper, turpis arcu iaculis neque, in sagittis libero leo ut sapien. Maecenas maximus pharetra magna non molestie. Mauris euismod fermentum varius. Donec pulvinar sit amet lorem at maximus. Donec vestibulum nibh eu varius tincidunt.\n",
        "                            ''', justified_paragraph))\n",
        " \n",
        " \n",
        "    story.append(PageBreak())\n",
        " \n",
        " \n",
        "    # Iterate through each condition to add it to the document\n",
        "    for condition, variants in result.items():\n",
        "        # Add condition name as a heading\n",
        "        story.append(Paragraph(condition, styles['Heading3']))\n",
        "        story.append(Paragraph(\"<br/>\", styles['Normal']))  # Add some spacing\n",
        "        story.append(Paragraph('Variant Classification and Significance', styles['Heading4']))\n",
        " \n",
        "        # Table data including the headers\n",
        "        table_data1 = [[\n",
        "            \"Chromosome\", \"Position\", \"Reference \\nAllele\", \"Alternate \\nAllele\",\n",
        "            \"Variant \\nQuality\", \"Filter \\nStatus\", \"ACMG \\nClassification\", \"ClinVar \\nSignificance\",\n",
        "        ]]\n",
        " \n",
        "        # Add the details for each variant\n",
        "        for variant in variants:\n",
        "            row = [\n",
        "                variant.get(\"Chromosome\", \"\"),\n",
        "                str(variant.get(\"Position\", \"\")),\n",
        "                variant.get(\"Reference Allele\", \"\"),\n",
        "                variant.get(\"Alternate Allele\", \"\"),\n",
        "                str(variant.get(\"Variant Quality\", \"\")),\n",
        "                variant.get(\"Filter Status\", \"\"),\n",
        "                variant.get(\"ACMG Classification\", \"\"),\n",
        "                variant.get(\"ClinVar Significance\", \"\")\n",
        "            ]\n",
        "            table_data1.append(row)\n",
        " \n",
        "        table1 = Table(table_data1, repeatRows=1) # colWidths=colWidths,\n",
        "        table1.setStyle(TableStyle([\n",
        "            ('GRID', (0,0), (-1,-1), 1, colors.black),\n",
        "            ('FONTNAME', (0,0), (-1,0), 'Helvetica-Bold'),\n",
        "            ('ALIGN', (0,0), (-1,-1), 'CENTER'),\n",
        "            ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),\n",
        "            ('BACKGROUND', (0,0), (-1,0), colors.lightgrey)\n",
        "        ]))\n",
        "       \n",
        "        story.append(table1)\n",
        " \n",
        "        story.append(Paragraph(\"<br/>\", styles['Normal']))  # Add some spacing\n",
        "        story.append(Paragraph('Variant Population Frequency', styles['Heading4']))\n",
        "        # Table data including the headers\n",
        "        table_data2 = [[\n",
        "            \"Chromosome\", \"Position\", \"Reference \\nAllele\", \"Alternate \\nAllele\",\n",
        "            \"Variant \\nQuality\", \"Filter \\nStatus\", \"gnomAD \\nAF\", \"1000g \\nAF\"\n",
        "        ]]\n",
        " \n",
        "        # Add the details for each variant\n",
        "        for variant in variants:\n",
        "            row = [\n",
        "                variant.get(\"Chromosome\", \"\"),\n",
        "                str(variant.get(\"Position\", \"\")),\n",
        "                variant.get(\"Reference Allele\", \"\"),\n",
        "                variant.get(\"Alternate Allele\", \"\"),\n",
        "                str(variant.get(\"Variant Quality\", \"\")),\n",
        "                variant.get(\"Filter Status\", \"\"),\n",
        "                str(variant.get(\"gnomAD AF\", \"\")),\n",
        "                str(variant.get(\"1000g AF\", \"\"))\n",
        "            ]\n",
        "            table_data2.append(row)\n",
        " \n",
        "        table2 = Table(table_data2, repeatRows=1) # colWidths=colWidths,\n",
        "        table2.setStyle(TableStyle([\n",
        "            ('GRID', (0,0), (-1,-1), 1, colors.black),\n",
        "            ('FONTNAME', (0,0), (-1,0), 'Helvetica-Bold'),\n",
        "            ('ALIGN', (0,0), (-1,-1), 'CENTER'),\n",
        "            ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),\n",
        "            ('BACKGROUND', (0,0), (-1,0), colors.lightgrey)\n",
        "        ]))\n",
        "       \n",
        "        story.append(table2)\n",
        "        story.append(PageBreak())\n",
        " \n",
        "        story.append(Paragraph(condition, styles['Heading3']))\n",
        "        story.append(Paragraph(\"<br/>\", styles['Normal']))  # Add some spacing\n",
        "        story.append(Paragraph('Variant Consequence and Inheritance Pattern', styles['Heading4']))\n",
        " \n",
        "        # Table data including the headers\n",
        "        table_data3 = [[\n",
        "            \"Chromosome\", \"Position\",\n",
        "            \"Reference \\nAllele\", \"Alternate \\nAllele\",\n",
        "            \"Consequence\", \"Inheritance \\nPattern\"\n",
        "        ]]\n",
        " \n",
        "        # Add the details for each variant\n",
        "        for variant in variants:\n",
        "            row = [\n",
        "                variant.get(\"Chromosome\", \"\"),\n",
        "                str(variant.get(\"Position\", \"\")),\n",
        "                variant.get(\"Reference Allele\", \"\"),\n",
        "                variant.get(\"Alternate Allele\", \"\"),\n",
        "                variant.get(\"Consequence\", \"\"),\n",
        "                variant.get(\"Inheritance Pattern\", \"\"),\n",
        "            ]\n",
        "            table_data3.append(row)\n",
        " \n",
        "        # Create the table for the current condition\n",
        "        #colWidths = [50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50]\n",
        "        table3 = Table(table_data3, repeatRows=1) # colWidths=colWidths,\n",
        "        table3.setStyle(TableStyle([\n",
        "            ('GRID', (0,0), (-1,-1), 1, colors.black),\n",
        "            ('FONTNAME', (0,0), (-1,0), 'Helvetica-Bold'),\n",
        "            ('ALIGN', (0,0), (-1,-1), 'CENTER'),\n",
        "            ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),\n",
        "            ('BACKGROUND', (0,0), (-1,0), colors.lightgrey)\n",
        "        ]))\n",
        "       \n",
        "        story.append(table3)\n",
        "        story.append(PageBreak())\n",
        " \n",
        "        story.append(Paragraph(condition, styles['Heading3']))\n",
        "        story.append(Paragraph(\"<br/>\", styles['Normal']))  # Add some spacing\n",
        "        story.append(Paragraph('Variant Age of Onset and Prevalence', styles['Heading4']))\n",
        " \n",
        "        # Table data including the headers\n",
        "        table_data4 = [[\n",
        "            \"Chromosome\", \"Position\",\n",
        "            \"Reference \\nAllele\", \"Alternate \\nAllele\",\n",
        "            \"Age of \\nOnset\",\"Prevalence\"\n",
        "        ]]\n",
        " \n",
        "        # Add the details for each variant\n",
        "        for variant in variants:\n",
        "            row = [\n",
        "                variant.get(\"Chromosome\", \"\"),\n",
        "                str(variant.get(\"Position\", \"\")),\n",
        "                variant.get(\"Reference Allele\", \"\"),\n",
        "                variant.get(\"Alternate Allele\", \"\"),\n",
        "                variant.get(\"Age of Onset\", \"\"),\n",
        "                variant.get(\"Prevalence\", \"\"),\n",
        "            ]\n",
        "            table_data4.append(row)\n",
        " \n",
        "        # Create the table for the current condition\n",
        "        #colWidths = [50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50, 50]\n",
        "        table4 = Table(table_data4, repeatRows=1) # colWidths=colWidths,\n",
        "        table4.setStyle(TableStyle([\n",
        "            ('GRID', (0,0), (-1,-1), 1, colors.black),\n",
        "            ('FONTNAME', (0,0), (-1,0), 'Helvetica-Bold'),\n",
        "            ('ALIGN', (0,0), (-1,-1), 'CENTER'),\n",
        "            ('VALIGN', (0,0), (-1,-1), 'MIDDLE'),\n",
        "            ('BACKGROUND', (0,0), (-1,0), colors.lightgrey)\n",
        "        ]))\n",
        "       \n",
        "        story.append(table4)\n",
        " \n",
        " \n",
        "        # Add a page break after each condition, except the last\n",
        "        if condition != list(result.keys())[-1]:\n",
        "            story.append(PageBreak())\n",
        " \n",
        " \n",
        "    # Build the PDF\n",
        "    doc.build(story, onFirstPage=myFirstPage, onLaterPages=myLaterPages)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "metadata": {},
      "outputs": [],
      "source": [
        "workdir = 'c:\\\\Users\\\\MSI\\\\Downloads'\n",
        "metadata_path = os.path.join(workdir,'patient_data_demo.json')\n",
        "result_path = os.path.join(workdir,'report_demo.json')\n",
        "\n",
        "pdf_file_path = os.path.join(workdir,'final_report_demo.pdf')\n",
        "\n",
        "json_to_pdf(metadata_path, result_path, pdf_file_path)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.8"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
